fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -0,0 +1,407 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.directory;
+
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.Tuple;
+import com.google.common.base.Verify;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.Lock;
+import org.apache.lucene.store.LockFactory;
+import org.apache.lucene.store.NoLockFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import java.io.IOException;
+import java.nio.file.NoSuchFileException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Directory implementation backed by FDB which attempts to
+ * model a file system on top of FoundationDB.
+ *
+ * A few interesting details:
+ *
+ * <ul>
+ *     <li>Each segment written in Lucene once flushed is <b>immutable</b>.</li>
+ *     <li>As more data is written into Lucene, files are continually merged via a <a href="https://lucene.apache.org/core/7_6_0/core/org/apache/lucene/index/MergePolicy.html">MergePolicy</a></li>
+ * </ul>
+ *
+ * @see <a href="https://lucene.apache.org/core/7_6_0/core/org/apache/lucene/store/Directory.html">Directory</a>
+ */
+@API(API.Status.EXPERIMENTAL)
+@NotThreadSafe
+public class FDBDirectory extends Directory {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FDBDirectory.class);
+    public static final int DEFAULT_BLOCK_SIZE = 16_384;
+    private final AtomicLong nextTempFileCounter = new AtomicLong();
+    private final Transaction txn;
+    private final Subspace subspace;
+    private final Subspace metaSubspace;
+    private final Subspace dataSubspace;
+    private final byte[] sequenceSubspaceKey;
+
+    private final LockFactory lockFactory;
+    private final int blockSize;
+    private final Cache<String, FDBLuceneFileReference> fileReferenceCache;
+    private final Cache<Pair<Long, Integer>, CompletableFuture<byte[]>> blockCache;
+    private final Map<String, Long> reads;
+
+    public FDBDirectory(@Nonnull Subspace subspace, @Nonnull Transaction txn) {
```

Comment:
Followup: these should probably be using `FDBRecordContext` to be consistent with the rest of the record layer.